spec.html

```diff
@@ -804,7 +804,6 @@ <h1>ReturnIfAbrupt</h1>
         <emu-alg>
           1. Let _hygienicTemp_ be AbstractOperation().
           1. If _hygienicTemp_ is an abrupt completion, return _hygienicTemp_.
-          1. Else if _hygienicTemp_ is a Completion Record, set _hygienicTemp_ to _hygienicTemp_.[[Value]].
```

Comment:
I’m not sure this is accurate; the original algorithm explicitly unwrapped the completion record. See the lines below (817)

Reply:
Yeah, but in this case, when the result of AbstractOperation() isn't going anywhere (isn't being bound to an alias), the "unwrapping" has no discernible effect.

Reply:
it is, though - `? Foo()` ends up becoming the result of `hygienicTemp`, as if this was a function and there was an implicit `return hygienicTemp`. Perhaps there's better wording we could use here to make that clear?

Reply:
This is the expansion of the shorthand step
```
1. ReturnIfAbrupt(AbstractOperation()).
```
so the result of AbstractOperation(), if it isn't abrupt, is discarded.

It sounds like you're thinking of a step like
```
1. Let _result_ be ReturnIfAbrupt(AbstractOperation()).
```

Reply:
@ljharb Note that this expansion is written in terms of a whole algorithm step, not any part of any algorithm step.

Reply:
ah, i misunderstood. so this isn’t for ?, its for ReturnIfAbrupt?

Reply:
Right. (The expansions for '?' just replace it with `ReturnIfAbrupt`, so don't get into temporaries or unwrapping.)

Reply:
So then wouldn’t this change the resulting inlining of `?`?

Reply:
This is only the case where the value is not used. I think you need to take a look at this section again. The rewrite rules for ReturnIfAbrupt are defined as pattern matches on full algorithm steps, and the rewrite I've changed only applies when the whole step is a call to ReturnIfAbrupt with its argument being an abstract operation. Nobody can be relying on the unobservable setting of a generated temporary variable.

Reply:
It would only affect those occurrences of `?` that expand to a step that says (or is otherwise equivalent to):
```
1. ReturnIfAbrupt(AbstractOperation()).
```
Literally, this would be occurrences of the form:
```
1. ? AbstractOperation()
```
But we have no steps that begin with `?` (outside of the `ReturnIfAbrupt Shorthands` clause).

However, we do have steps of the form:
```
1. Perform ? Foo(...).
```
which presumably expands to
```
1. Perform ReturnIfAbrupt(Foo(...)).
```
which you could argue is "otherwise equivalent" to:
```
1. ReturnIfAbrupt(Foo(...)).
```
whose expansion **is** altered by the proposed change.

But that alteration doesn't change semantics, because, in the original, the result of Foo(), if it isn't abrupt, is discarded anyway.

Reply:
Thanks for explaining. Although it sounds like there might be more refinement possible here, the things I’m concerned about don’t apply, so this change seems good to me!